### ISSUE TITLE
Solver failing

### ISSUE THREAD

**Issue by alanrandrade** [TorXakis/TorXakis#795]
Hi,

I created a simple model, which makes a query and expects a return.

The query is defined in `TYPEDEF Actions ::= Query {in :: ParameterType; arg :: Int} ENDDEF`
The response is defined in: 
```
TYPEDEF ServerActions
    ::=
        ReturnQuery {
            out :: ParameterType; 
            myList :: ListInt;
            error :: EResponseType}
ENDDEF
```

Given a query:
- In case the attribute `arg` has a value between 1 and 15, the attribute `error` (EResponseType) in the return should be `SUCCESS` and the list of integers should have at least one element.
- If the attribute `arg` is 'invalid', we expect a `FAILURE` and the list of integers should have no elements.

This behaviour is expressed in the following model:
```

STAUTDEF mySTAUTDEF[ Input::Actions; Output::ServerActions] () ::=
        STATE  state0, state1, state2

        INIT   state0

        TRANS   --if a valid attribute arg is given as input, we go to state 1, where we expect a  SUCCESS error.
                state0  -> Input ? a [[isValidArg(arg(a))]] 
                                    -> state1

                --if an invalid attribute arg is given as input, we go to state 2, where we expect FAILURE error.
                state0  -> Input ? a [[not(isValidArg(arg(a)))]] 
                                    -> state2

                --  If it is a SUCCESS, the list should have at least one element
                state1  -> Output ? b [[isSUCCESS(error(b)) /\
                                        (lengthList(myList(b)) >= 1)
                                        ]] 
                                        -> state0

                --  If it is a FAILURE, the list should be empty
                state2  -> Output ? b [[isFAILURE(error(b)) /\
                                        (lengthList(myList(b)) == 0)
                                        ]]
                                        -> state0
                
ENDDEF
```

The problem is that when I run this model in tester mode and use nc -l 7890 to send the responses, I get _unfoldCTbranch: Solve Unknown_ messages in some cases .

E.g

> TXS >>  .....5: IN: Act { { ( Input, [ Query(A,16) ] ) } }
> TXS >>  unfoldCTbranch: Solve Unknown
> TXS >>  .....6: OUT: Act { { ( Output, [ ReturnQuery(A,NilInt,FAILURE) ] ) } }

Am I missing something? I've added the full .txs file (in .txt, because git does not allow .txs) and a log containing a sequence of interactions.

[simpleListInt.txt](https://github.com/TorXakis/TorXakis/files/2470496/simpleListInt.txt)
[final-log.txt](https://github.com/TorXakis/TorXakis/files/2470497/final-log.txt)

Thank you in advance,
Alan

**Comment by pjljvandelaar** [comment]
A few questions
1. Is the same behaviour also observed with the latest release of z3?
2. What version of TorXakis are you using?
3. Is the log file made with the model? I see no constraints between the ParameterType of the query and the returnQuery. Yet the ParameterType of the returnQuery always matches that of the query. I did not expect that with the default random generation.

**Comment by alanrandrade** [comment]
1. Yes, the behaviour was the same with z3 version 4.8.0 (see log)
2. I am using the 0.6.0 version (Linux) in a Ubutu 16.04
3. Yes. I ran it in _tester mode_ and sent the responses manually, that is the reason why they match even though there are no constraints. (ran it again without caring about matching ParameterType, see log-error.txt)

I believe I found the cause of the problem, which seems to me like a small bug.

After replacing `(lengthList(myList(b)) == 0)` with `(isNilInt(myList(b)))` there are no _unfoldCTbranch: Solve Unknown_ messages. 

If you run the model in stepper mode using `(lengthList(myList(b)) == 0)` it will stop for a few minutes in a few cases before resulting in _unfoldCTbranch: Solve Unknown_, with `(isNilInt(myList(b)))` everything goes the way it should be (smooth-log.txt)

```
state2  -> Output ? b [[isFAILURE(error(b)) /\
                                        --this breaks it
                                        --(lengthList(myList(b)) == 0)
                                        (isNilInt(myList(b)))
                                        ]]
                                        -> state0
```


[simpleListInt.txt](https://github.com/TorXakis/TorXakis/files/2474388/simpleListInt.txt)
[log-error.txt](https://github.com/TorXakis/TorXakis/files/2474393/log-error.txt)
[smooth-log.txt](https://github.com/TorXakis/TorXakis/files/2474397/smooth-log.txt)

**Comment by pjljvandelaar** [comment]
Indeed it is smarter to use isNilInt for the empty list and isCstrInt for the non-empty list. The problemsolver in  those cases doesn't have to solve a problem that includes the (recursive) length function.

**Comment by pjljvandelaar** [comment]
After my holidays, i will look into the issue and report back my findings.

**Comment by pjljvandelaar** [comment]
Could reproduce issue and isolate it. Send bug report to Z3:  https://github.com/Z3Prover/z3/issues/1895

**Comment by pjljvandelaar** [comment]
I could reproduce the issue and isolate it.
I submitted a bug report to Z3: https://github.com/Z3Prover/z3/issues/1895

**Comment by pjljvandelaar** [comment]
Note that the original problem, leads in stepper and simulator mode to a query that is too hard to decide for current problem solvers:

The statement `lengthList(myList(b)) == 0` has a solution.
TorXakis however wants to generate a random solution when possible, 
so it will ask for a solution to the two statements 
`lengthList(myList(b)) == 0` and `not (isNilInt(myList(b)))`, 
to find out whether multiple solutions are possible.
However to prove that this query has no solutions, requires one to use the monotonicity of the lengthList function (i.e., the longer the list the large the length value).
To my knowledge both z3 and cv4 lack this capability.

The already discussed alternative problem that uses `isNilInt` and `isCstrInt` does not create a query that can't be solved by current problem solvers.

**Comment by alanrandrade** [comment]
Thank you for looking into this, @pjljvandelaar. I will stick to `isNilInt` for now.

Another thing I observed with this small model is the way the test cases are generated. 

- When IncrementChoice is used as `param_Randomization`, diverse `ParameterTypes` (A or B) are used, `arg` has random values, but more diversity of valid arguments `( (arg >= 1) /\ (arg <= 15)  )` would be ideal, instead of only the case 1 (see log-error.txt (tester mode) from previous comment or IncrementChoice-log-stepper-mode.txt). 

I quickly discussed about this with @tretmans and he advised me to look into TrueBins and try to play around with its parameters to get a better diversity of values for test cases.

- When TrueBins is used, diverse valid/invalid args (-73, 1, 3, ... see logs) are generated, but B is always used as a ParameterType for all test cases. I played around with the numbers of bins, but did not succeed in getting random values in a more optimal way (diverse ParameterTypes (A and B) and valid/invalid args).

I understand that you are currently working on ways to improve randomization, I am just wondering if there is a currently available way on how to improve this that I don't know about and could use.

[IncrementChoice-log-stepper-mode.txt.txt](https://github.com/TorXakis/TorXakis/files/2505407/IncrementChoice-log-stepper-mode.txt.txt)
[TrueBins-log-stepper-mode.txt.txt](https://github.com/TorXakis/TorXakis/files/2505409/TrueBins-log-stepper-mode.txt.txt)
[TrueBins-log-tester-mode.txt](https://github.com/TorXakis/TorXakis/files/2505412/TrueBins-log-tester-mode.txt)

Thank you in advance,
Alan

**Comment by pjljvandelaar** [comment]
Please use IncrementBins. It should be the default. Maybe reduce the integer range, since you are not interested in overflow ( numbers larger than 2^64)

**Comment by pjljvandelaar** [comment]
With the following similar program (growingList.txs)
```
TYPEDEF List ::= Nil
               | Cstr{head::Int; tail::List}
ENDDEF

CHANDEF Chans ::=
    Input     ;
    Output    :: List
ENDDEF

PROCDEF proc [Input; Output :: List](l :: List) ::=
        Input
    >-> Output ? n [[ IF isCstr(n) THEN tail(n) == l ELSE False FI ]]
    >-> proc [Input, Output] (n)
ENDDEF

MODELDEF Model ::=
    CHAN IN    Input
    CHAN OUT   Output

    BEHAVIOUR  
        proc [Input,Output] (Nil)
ENDDEF
```
I get 
```
PS C:\experiment> torxakis.exe .\growingList.txs

TXS >>  TorXakis :: Model-Based Testing

TXS >>  txsserver starting: "NLYBSTQVP4NB77Y.code1.emi.philips.com" : 61094
TXS >>  Solver "z3" initialized : Z3 [4.8.0 - build hashcode a12af4a619a0]
TXS >>  TxsCore initialized
TXS >>  input files parsed:
TXS >>  .\growingList.txs
TXS >> stepper Model
TXS >>  Stepper started
TXS >> step 10
TXS >>  .....1: Act { { ( Input, [] ) } }
TXS >>  .....2: Act { { ( Output, [ Cstr(-86,Nil) ] ) } }
TXS >>  .....3: Act { { ( Input, [] ) } }
TXS >>  .....4: Act { { ( Output, [ Cstr(-56,Cstr(-86,Nil)) ] ) } }
TXS >>  .....5: Act { { ( Input, [] ) } }
TXS >>  .....6: Act { { ( Output, [ Cstr(0,Cstr(-56,Cstr(-86,Nil))) ] ) } }
TXS >>  .....7: Act { { ( Input, [] ) } }
TXS >>  .....8: Act { { ( Output, [ Cstr(-90,Cstr(0,Cstr(-56,Cstr(-86,Nil)))) ] ) } }
TXS >>  .....9: Act { { ( Input, [] ) } }
TXS >>  ....10: Act { { ( Output, [ Cstr(50,Cstr(-90,Cstr(0,Cstr(-56,Cstr(-86,Nil))))) ] ) } }
TXS >>  PASS
```
I assume this is random enough for you?

Note my parameters are just the default values:
``` 
TXS >> param
TXS >>  param_ImpRel = IOCO
TXS >>  param_IncrementChoice_IntPower = 4
TXS >>  param_IncrementChoice_IntRange = 65536
TXS >>  param_IncrementChoice_MaxGeneratedStringLength = 10
TXS >>  param_InputCompletion = ANGELIC
TXS >>  param_RandSolve_IntHalf = 1000
TXS >>  param_RandSolve_IntNum = 5
TXS >>  param_Randomization = IncrementBins
TXS >>  param_Sim_inputEager = 0
TXS >>  param_Test_inputEager = 3
TXS >>  param_TrueBins_Next = Linear
TXS >>  param_TrueBins_NrOfBins = 10
TXS >>  param_TrueBins_StringLength = 6
TXS >>  param_TrueBins_StringMode = Regex
TXS >>  param_max_rand_depth = 4
TXS >>  param_Sim_deltaTime = 2000
TXS >>  param_Sim_ioTime = 10
TXS >>  param_Sut_deltaTime = 2000
TXS >>  param_Sut_ioTime = 10
```

**Comment by alanrandrade** [comment]
The problem was indeed with the randomization mode. 

I used the Linux installer torxakis_0.6.0_amd64.deb and for that reason I did not have the most up-to-date randomization options. I was having some problems to build torxakis in Ubuntu 16.04, but yesterday I put some extra effort into it after your response and managed to build it.

Now I am using the newest version of torxakis and IncrementBins randomization works beautifully.

Thank you again for your help! :)
